lemur/plugins/lemur_gcp/plugin.py

```diff
@@ -48,81 +44,108 @@ class GCPDestinationPlugin(DestinationPlugin):
     ]
 
     def upload(self, name, body, private_key, cert_chain, options, **kwargs):
-
         try:
             ssl_certificate_body = {
-                "name": self._certificate_name(body),
-                "certificate": self._full_ca(body, cert_chain),
+                "name": certificates.get_name(body),
+                "certificate": certificates.full_ca(body, cert_chain),
                 "description": "",
                 "private_key": private_key,
             }
-            credentials = self._get_gcp_credentials(options)
-            return self._insert_gcp_certificate(
+            credentials = auth.get_gcp_credentials(self, options)
+            return certificates.insert_certificate(
                 self.get_option("projectID", options),
                 ssl_certificate_body,
                 credentials,
             )
-
+        except exceptions.AlreadyExists:
```

Comment:
Im not sure I understand the value in this exception block. If I understand correctly the aim is to "no-op" an `AlreadyExists` Error. I think this is very valid error that we want to log and then raise to our user (I've certainly seen come up a few times in testing).
The value in the exception block below is that we catch whichever error that may come up in the execution flow, log it in a format that we want and then re-raise the error so the frontend will display it to the user.

Reply:
There's a codepath where lemur tries to upload the same certificate to a destination. I don't think that should necessarily be an error - the AWS destination plugin also ignores when a cert already exists.

Reply:
I think the main scenario where ignoring this error could cause problems is if a certificate is renewed and uploaded to GCP with the same name as the certificate it is replacing. I think we would have to make sure the issue date is never truncated to guarantee uniqueness if not already the case...

Reply:
We can check if GCP de-dupes on name or name + body + chain as a first step. And for the naming convention, we could change it so:
`len(truncate(f"ssl-{cn}-{authority}") + -f"{issued_on}" == 63`
Any objection to doing that?

Reply:
We might be forced to use the certificate serial number here instead of the issue date because otherwise a user won't be able re-issue and upload certificate from the same CA on the same day. (Say you choose the wrong key length for example.)

Reply:
Good point - based on some reading though the serial number is only [unique](https://stackoverflow.com/questions/9104108/is-serial-number-a-unique-key-for-x509-certificate) to the CA and I believe there are some plans to add another CA later. Even though a collision is probably unlikely for setting the name to be the cert number, I will go ahead with using the sha256 fingerprint of the cert which should be unique across different CAs.

Reply:
The name includes the CA which should prevent that collision, but sure the fingerprint works too.

Reply:
Done - changed the name to be `ssl-{fingerprint}`.

Reply:
Personally, I would prefer to have a more human readable name for the GCP certificates. Given that we know our constraint is 63 chars how would we feel about the name being something such as.
`common_name[:30]-issuer[:10]-serial[:10]-issued_on[:10]`
Im not tied to the convention above, but I would like something that sticks more or less to format we currently have while still being unique. 

I think its important we all agree on the naming convention before moving forward because the name will be used both when we upload the cert as well as when we update the LB. So once we choose the naming convention we are pretty tied to it.

Reply:
Discussed offline with Mitch, going to change the name to be: `{cn}-{issuer}-{serial}`. The CN is trimmed so that the len(full name) <= 63 and the full issuer + serial should make it unique (for the scenario where we create a create a cert with the wrong encryption method and need to recreate it on the same day).